Added deserialisation tests for length-of Plutus ByteStrings > 64.

TimSheard opened the pull request:
In order to validate data in both AuxiliaryData and in the WitnessSet, deserializing
a Plutus Data, with a ByteString type, whose length is greater than 64 now raises
a deserialisation error. Added a new combinator (<?) to Data.Coders that lets any
'constructor' raise an error. Also added roundtrip tests to be sure we actually
catch these cases.

TimSheard committed:
Check length plutus bytestring data, and double check some tests.

In order to validate data in both AuxiliaryData and in the WitnessSet, deserializing
a Plutus Data, with a ByteString type, whose length is greater than 64 now raises
a deserialisation error. Added a new combinator (<?) to Data.Coders that lets any
'constructor' raise an error. Also added roundtrip tests to be sure we actually
catch these cases.

Before running plutus scripts, we collect their input data using their hashes and the Witnesses.
Previous PredicateFailure tests should ensure we find Data for every script, BUT
the consequences of not finding Data means scripts can get dropped, so things
might validate that shouldn't. So we double check that every Script has its Data, and
if that is not the case, a PredicateFailure is raised in the Utxos rule.

JaredCorduan reviewed (approved):
If we can add the `FromCBOR (Annotator (Data era))` back in, I think it looks great!

nc6 reviewed (changes requested):
Shouldn't add additional consttaints on the API types, otherwise looks fine.

TimSheard commented:
I cleaned this up, addressed the (Show(CompactAddr era)) comments from Nick, rebased on the most recent master.
I am still getting a serialisation error in the benchmarks.

bench: Failed to decode state: DecoderErrorDeserialiseFailure "state" (DeserialiseFailure 6689 "expected integer")
CallStack (from HasCallStack):
  error, called at bench/Bench/Cardano/Ledger/ApplyTx.hs:100:25 in main:Bench.Cardano.Ledger.ApplyTx
Benchmark bench: ERROR
cabal: Benchmarks failed for bench:bench from cardano-ledger-test-0.1.0.0.
